python/examples/math/program.py

```diff
@@ -0,0 +1,166 @@
+from __future__ import annotations
+import asyncio
+import json
+from textwrap import dedent, indent
+from typing import Mapping, TypeVar, Any, Callable, Awaitable, TypedDict, Annotated,  NotRequired, override
+
+from typechat import (
+    Failure,
+    Result,
+    Success,
+    TypeChatModel,
+    TypeChatValidator,
+    TypeChatTranslator,
+    python_type_to_typescript_schema,
+)
+import collections.abc
+
+T = TypeVar("T", covariant=True)
+
+def Doc(s: str) -> str:
+    return s
+
+
+program_schema_text = '''
+// A program consists of a sequence of function calls that are evaluated in order.
+export type Program = {
+    "@steps": FunctionCall[];
+}
+
+// A function call specifies a function name and a list of argument expressions. Arguments may contain
+// nested function calls and result references.
+export type FunctionCall = {
+    // Name of the function
+    "@func": string;
+    // Arguments for the function, if any
+    "@args"?: Expression[];
+};
+
+// An expression is a JSON value, a function call, or a reference to the result of a preceding expression.
+export type Expression = JsonValue | FunctionCall | ResultReference;
+
+// A JSON value is a string, a number, a boolean, null, an object, or an array. Function calls and result
+// references can be nested in objects and arrays.
+export type JsonValue = string | number | boolean | null | { [x: string]: Expression } | Expression[];
+
+// A result reference represents the value of an expression from a preceding step.
+export type ResultReference = {
+    // Index of the previous expression in the "@steps" array
+    "@ref": number;
+};
+'''
+
+
+ResultReference = TypedDict(
+    "ResultReference", {"@ref": Annotated[int, Doc("Index of the previous expression in the 'steps' array")]}
+)
+
+FunctionCall = TypedDict(
+    "FunctionCall",
+    {
+        "@func": Annotated[str, Doc("Name of the function")],
+        "@args": NotRequired[Annotated[list["Expression"], Doc("Arguments for the function, if any")]],
+    },
+)
+
+JsonValue = str | int | float | bool | None | dict[str, "Expression"] | list["Expression"]
+Expression = JsonValue | FunctionCall | ResultReference
+
+JsonProgram = TypedDict("Program", {"@steps": list[FunctionCall]})
+
+
+async def evaluate_json_program(program: Mapping[str, Any], onCall: Callable[[str, list[Any]], Awaitable[Any]]) -> Any:
```

Comment:
Can we use Dict[str, Any], I haven't seen Mapping being used as frequently?
Instead of list[Any] can we Change it to **List**[any]. Also for clarity, can we use type hints from the typing module like List, Dict instead of list, dict etc.

Reply:
As it turns out, the typing module's Dict, List, Set, Tuple, Type and other types are deprecated aliases for their lowercase syblings (https://docs.python.org/3/library/typing.html#deprecated-aliases). 

The Mapping va. Dict difference is also interesting. The Mapping definition is more generic (just needs a type to define __getitem__,__len__,__iter__) from the ABCs (https://docs.python.org/3/library/collections.abc.html#collections-abstract-base-classes) while the dict type is a more concrete implememtation. The notes I've seen online recommend using a Mapping type unless you strictly need a dict type.